homeassistant/components/soma/cover.py

```diff
@@ -0,0 +1,81 @@
+"""
+Support for Soma Covers.
+
+For more details about this platform, please refer to the documentation at
+"""
+
+import logging
+
+from homeassistant.components.cover import CoverDevice, ATTR_POSITION
+from homeassistant.components.soma import DOMAIN, SomaEntity, DEVICES, API
+from homeassistant.exceptions import PlatformNotReady
+
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up the Soma cover platform."""
+    def get_covers():
+        """Retrieve covers."""
+        devices = hass.data[DOMAIN][DEVICES]
+
+        return [SomaCover(cover, hass.data[DOMAIN][API]) for cover in
+                devices]
+
+    async_add_entities(await hass.async_add_executor_job(get_covers), True)
+
+
+async def async_setup_platform(hass, config, async_add_entities,
+                               discovery_info=None):
+    """Old way of setting up platform.
+
+    Can only be called when a user accidentally mentions the platform in their
+    config. But even in that case it would have been ignored.
+    """
+    pass
+
+
+class SomaCover(SomaEntity, CoverDevice):
+    """Representation of a Soma cover device."""
+
+    async def async_update(self):
+        """Update the device with the latest data."""
+        await super().async_update()
+
+    def close_cover(self, **kwargs):
+        """Close the cover."""
+        if self.api.close_shade(self.device['mac'])['result'] != 'success':
+            raise PlatformNotReady()
+
+    def open_cover(self, **kwargs):
+        """Open the cover."""
+        if self.api.open_shade(self.device['mac'])['result'] != 'success':
+            raise PlatformNotReady()
+
+    def stop_cover(self, **kwargs):
+        """Stop the cover."""
+        # Set cover position to some value where up/down are both enabled
+        self.current_position = 50
+        if self.api.stop_shade(self.device['mac'])['result'] != 'success':
+            raise PlatformNotReady()
+
+    def set_cover_position(self, **kwargs):
+        """Move the cover shutter to a specific position."""
+        self.current_position = kwargs[ATTR_POSITION]
+        if self.api.set_shade_position(
+                self.device['mac'], kwargs[ATTR_POSITION])['result'] != 'success':
+            raise PlatformNotReady()
+
+    @property
+    def current_cover_position(self):
+        """Return the current position of cover shutter."""
+        return self.current_position
+
+    @property
+    def is_closed(self):
+        """Return if the cover is closed."""
+        is_closed = False
+        if self.current_position == 100:
+            is_closed = True
```

Comment:
Home Assistant has the opposite interpretation of position and open.
> None is unknown, 0 is closed, 100 is fully open.

We need to report the correct state according to Home Assistant standards, no matter what the device api convention is. So either invert here or in `current_cover_position` property.

Reply:
Fixed this.

Reply:
I think we have to change how we report current position too, and do `100 - self.current_position` for `current_cover_position`. Right? Otherwise there will be a mismatch what the decide thinks is closed and what home assistant thinks is closed. Sorry for my confusion.

Reply:
I can change this in the API layer in the hub. I already have that functionality in the hub because HomeKit and the native direction of the device are also swapped.

Reply:
Ok, good!